# yelp capstone project
Consist of 5 tables e.g Business, Review, User, Tip, Checkin.\
For dataset details kindly [visit](https://www.yelp.com/dataset/documentation/main)\
Bootcamper : Robert, Sining, Theresa, Kamruddin

The project was initialized and capstoned to pass the knowledge and learning into application.

Now we are importing necessary packages 

In [1]:
import json
import pandas as pd

In [ ]:
# add 'data/' to .gitignore so that you do not upload data to github
!echo 'data/' >> .gitignore

Opening json file from business dataset together with unpacking 'arrributes' columns to find to particular hypothetic columns.

In [2]:
data = []

with open('data/yelp_academic_dataset_business.json') as f:

    for line in f:
        
        #Parse the JSON data
        json_dict = json.loads(line)

        if json_dict['attributes']:
            if 'BusinessParking' in json_dict['attributes'].keys():

                json_dict['attributes']['BusinessParking'] = eval(json_dict['attributes']['BusinessParking'])

        data.append(json_dict)

normlize dataset

In [3]:
biz_df = pd.json_normalize(data)

In [5]:
biz_df.shape

(150346, 65)

Now we have 150,346 rows and 65 columns. We need to clean on column level then deep dive into cells.

## Dropping empty category rows and reset index on existing dataframe

In [6]:
# drop row that have no category labels
biz_df.dropna(subset='categories', inplace=True)
# reset index
biz_df.reset_index(drop=True, inplace=True)

In [7]:
# import category labels from CSV and filter dataframe to keep row that have category label Approx 30 min
cat_data = pd.read_csv('data/categories.csv', header=None)
categories = cat_data[0].tolist()
filtered_df = pd.DataFrame()

for index, item in biz_df['categories'].items():

    for category in categories:
        if category in item:
            filtered_df = pd.concat([filtered_df, biz_df[index:index+1]])
            break

In [8]:
filtered_df.columns

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'stars', 'review_count', 'is_open',
       'categories', 'hours', 'attributes.ByAppointmentOnly',
       'attributes.BusinessAcceptsCreditCards', 'hours.Monday',
       'hours.Tuesday', 'hours.Wednesday', 'hours.Thursday', 'hours.Friday',
       'hours.Saturday', 'attributes.BikeParking',
       'attributes.RestaurantsPriceRange2', 'attributes.CoatCheck',
       'attributes.RestaurantsTakeOut', 'attributes.RestaurantsDelivery',
       'attributes.Caters', 'attributes.WiFi',
       'attributes.BusinessParking.garage',
       'attributes.BusinessParking.street',
       'attributes.BusinessParking.validated',
       'attributes.BusinessParking.lot', 'attributes.BusinessParking.valet',
       'attributes.WheelchairAccessible', 'attributes.HappyHour',
       'attributes.OutdoorSeating', 'attributes.HasTV',
       'attributes.RestaurantsReservations', 'attributes.DogsAllowed',
       'hou

### Cleaning columns from business table

In [9]:
#1 dropping less important columns
filtered_df.drop(['hours','attributes.ByAppointmentOnly', 'attributes.BikeParking','attributes.CoatCheck',
       'attributes.Caters', 'attributes.WiFi', 'attributes.WheelchairAccessible', 'attributes.HappyHour', 'attributes.HasTV', 'attributes.DogsAllowed', 'attributes.GoodForKids',
       'attributes.RestaurantsAttire','attributes.NoiseLevel', 'attributes.GoodForMeal', 'attributes.BusinessAcceptsBitcoin', 'attributes.Smoking','attributes.Music',
       'attributes.GoodForDancing', 'attributes.AcceptsInsurance', 'attributes.BestNights', 'attributes.BYOB', 'attributes.Corkage', 'attributes.BYOBCorkage',
       'attributes.HairSpecializesIn','attributes.AgesAllowed','attributes.DietaryRestrictions','attributes.BusinessParking','attributes', 'attributes.Ambience'], axis=1, inplace=True)

In [10]:
filtered_df.columns

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'stars', 'review_count', 'is_open',
       'categories', 'attributes.BusinessAcceptsCreditCards', 'hours.Monday',
       'hours.Tuesday', 'hours.Wednesday', 'hours.Thursday', 'hours.Friday',
       'hours.Saturday', 'attributes.RestaurantsPriceRange2',
       'attributes.RestaurantsTakeOut', 'attributes.RestaurantsDelivery',
       'attributes.BusinessParking.garage',
       'attributes.BusinessParking.street',
       'attributes.BusinessParking.validated',
       'attributes.BusinessParking.lot', 'attributes.BusinessParking.valet',
       'attributes.OutdoorSeating', 'attributes.RestaurantsReservations',
       'hours.Sunday', 'attributes.Alcohol',
       'attributes.RestaurantsTableService',
       'attributes.RestaurantsGoodForGroups', 'attributes.DriveThru',
       'attributes.Open24Hours', 'attributes.RestaurantsCounterService'],
      dtype='object')

In [11]:
#2 renaming the header
filtered_df.rename(columns={'attributes.BusinessAcceptsCreditCards': 'accepts_credit_cards','hours.Monday':'monday', 'hours.Tuesday':'tuesday','hours.Wednesday':'wednesday',
                       'hours.Thursday':'thursday','hours.Friday':'friday','hours.Saturday':'saturday','hours.Sunday':'sunday','attributes.RestaurantsPriceRange2':'price_range',
                       'attributes.RestaurantsTakeOut':'take_out','attributes.RestaurantsDelivery':'delivery','attributes.BusinessParking.garage':'parking_garage',
                       'attributes.BusinessParking.street':'parking_street','attributes.BusinessParking.validated':'parking_validated','attributes.BusinessParking.lot':'parking_lot',
                       'attributes.BusinessParking.valet':'parking_valet','attributes.OutdoorSeating':'outdoor_seating','attributes.RestaurantsReservations':'seats_reservations',
                       'attributes.Alcohol':'alcohol','attributes.Ambience':'ambience','attributes.RestaurantsTableService':'table_service','attributes.RestaurantsGoodForGroups':'for_groups',
                       'attributes.DriveThru':'drive_thru','attributes.Open24Hours':'open_24_hours','attributes.RestaurantsCounterService':'counter_service'}, inplace=True)

In [12]:
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 66819 entries, 3 to 150237
Data columns (total 36 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   business_id           66819 non-null  object 
 1   name                  66819 non-null  object 
 2   address               66819 non-null  object 
 3   city                  66819 non-null  object 
 4   state                 66819 non-null  object 
 5   postal_code           66819 non-null  object 
 6   latitude              66819 non-null  float64
 7   longitude             66819 non-null  float64
 8   stars                 66819 non-null  float64
 9   review_count          66819 non-null  int64  
 10  is_open               66819 non-null  int64  
 11  categories            66819 non-null  object 
 12  accepts_credit_cards  58084 non-null  object 
 13  monday                49484 non-null  object 
 14  tuesday               52692 non-null  object 
 15  wednesday             5

In [13]:
filtered_df.columns

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'stars', 'review_count', 'is_open',
       'categories', 'accepts_credit_cards', 'monday', 'tuesday', 'wednesday',
       'thursday', 'friday', 'saturday', 'price_range', 'take_out', 'delivery',
       'parking_garage', 'parking_street', 'parking_validated', 'parking_lot',
       'parking_valet', 'outdoor_seating', 'seats_reservations', 'sunday',
       'alcohol', 'table_service', 'for_groups', 'drive_thru', 'open_24_hours',
       'counter_service'],
      dtype='object')

### Cleaning data inside the cells

In [14]:
filtered_df['alcohol'].unique()

array(["u'none'", nan, "u'full_bar'", "'none'", "'full_bar'",
       "u'beer_and_wine'", "'beer_and_wine'", 'None'], dtype=object)

In [15]:
filtered_df['alcohol'] = filtered_df['alcohol'].str.replace("u'", "").str.replace("'", "")

In [16]:
filtered_df['alcohol'].unique()

array(['none', nan, 'full_bar', 'beer_and_wine', 'None'], dtype=object)

### copying the dataframe to not loose the original df and find parking availability 

In [17]:
filter_df = filtered_df.copy()

In [18]:
# creating function to check the Values either true in any case
def parking_check(row):
    if row['parking_garage'] == True or row['parking_street'] == True or row['parking_validated'] == True or row['parking_lot'] == True or row['parking_valet'] == True:
        return True
    else:
        return False

In [19]:
# running parking check by row if anywhere found True returns True in a new Column else all returning False.
filter_df['parking_available'] = filter_df.apply(parking_check, axis=1)

In [20]:
# dropping all the parking extensions from the table to be in more cleaned way
filter_df.drop(['parking_garage', 'parking_street','parking_validated','parking_lot','parking_valet'], axis=1, inplace=True)

In [21]:
filter_df

business_id                           name  \
3       MTSW4McQd7CbVtyjqoe9mw             St Honore Pastries   
4       mWMc6_wTdE0EUBKIGXDVfA       Perkiomen Valley Brewery   
5       CF33F8-E6oudUQ46HnavjQ                 Sonic Drive-In   
8       k0hlBqXX-Bt0vf1op7Jr1w          Tsevi's Pub And Grill   
9       bBDDEgkFA1Otx9Lfe7BZUQ                 Sonic Drive-In   
...                        ...                            ...   
150224  cM6V90ExQD6KMSU3rRB5ZA              Dutch Bros Coffee   
150225  1jx1sfgjgVg0nM6n3p0xWA           Savaya Coffee Market   
150233  WnT9NIzQgLlILjPT0kEcsQ  Adelita Taqueria & Restaurant   
150236  2O2K6SXPWv56amqxCECd4w                   The Plum Pit   
150237  hn9Toz3s-Ei3uZPt7esExA          West Side Kebab House   

                      address          city state postal_code   latitude  \
3                 935 Race St  Philadelphia    PA       19107  39.955505   
4               101 Walnut St    Green Lane    PA       18054  40.338183   
5               615 S Main St  Ashland City    TN       37015  36.269593   
8           8025 Mackenzie Rd        Affton    MO       63123  38.565165   
9         2312 Dickerson Pike     Nashville    TN       37207  36.208102   
...                       ...           ...   ...         ...        ...   
150224    1181 N Milwaukee St         Boise    ID       83704  43.615401   
150225      11177 N Oracle Rd    Oro Valley    AZ       85737  32.409552   
150233          1108 S 9th St  Philadelphia    PA       19147  39.935982   
150236        4405 Pennell Rd         Aston    DE       19014  39.856185   
150237  2470 Guardian Road NW      Edmonton    AB     T5T 1K8  53.509649   

         longitude  stars  review_count  ...  outdoor_seating  \
3       -75.155564    4.0            80  ...            False   
4       -75.471659    4.5            13  ...              NaN   
5       -87.058943    2.0             6  ...             True   
8       -90.321087    3.0            19  ...             True   
9       -86.768170    1.5            10  ...             True   
...            ...    ...           ...  ...              ...   
150224 -116.284689    4.0            33  ...             True   
150225 -110.943073    4.5            41  ...             True   
150233  -75.158665    4.5            35  ...            False   
150236  -75.427725    4.5            14  ...              NaN   
150237 -113.675999    4.5            18  ...              NaN   

       seats_reservations     sunday   alcohol table_service for_groups  \
3                     NaN   7:0-21:0      none           NaN        NaN   
4                     NaN  12:0-18:0       NaN           NaN        NaN   
5                   False   8:0-22:0      none         False       True   
8                   False        NaN  full_bar           NaN       True   
9                   False   6:0-21:0      none         False      False   
...                   ...        ...       ...           ...        ...   
150224              False    0:0-0:0      none           NaN       True   
150225                NaN   6:0-15:0       NaN           NaN        NaN   
150233              False  11:0-22:0      none          True       True   
150236                NaN    0:0-0:0      none           NaN        NaN   
150237                NaN  11:0-22:0       NaN          True       True   

       drive_thru open_24_hours counter_service parking_available  
3             NaN           NaN             NaN              True  
4             NaN           NaN             NaN              True  
5            True           NaN             NaN             False  
8             NaN           NaN             NaN              True  
9            True           NaN             NaN             False  
...           ...           ...             ...               ...  
150224        NaN           NaN             NaN              True  
150225        NaN           NaN             NaN              True  
150233        NaN           NaN

### Now we have 66,819 rows and 32 analysable columns for further process.
### Calling review date and filter selected business_id

In [24]:
filtered = filter_df['business_id'].to_list()
import pandas as pd
column_names = ['review_id', 'user_id', 'business_id', 'stars', 'useful', 'funny', 'cool', 'text', 'date']
chunks = pd.read_json('data/y_review.json', lines=True, chunksize=10000)
with open('data/review_category_filt.csv', 'w') as f:
    header = ','.join(column_names)
    f.write(header + '\n')
for chunk in chunks:
    df = pd.DataFrame(chunk)
    for index, item in df['business_id'].items():
        if item in filtered:
            chunk[index:index+1].to_csv('data/review_category_filt.csv', mode='a', header=False, index=False)
    break

In [25]:
output = pd.read_csv('data/review_filtered.csv')
output.shape

(7938, 9)